src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/SuddenHighwayTypeChangeCheck.java

```diff
@@ -0,0 +1,248 @@
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Edge edgeBeingVerified = (Edge) object;
+
+        final List<Edge> completeWayEdges = new ArrayList<>(
+                new OsmWayWalker(edgeBeingVerified).collectEdges());
+
+        final Node firstEdgeStartNode = completeWayEdges.get(0).start();
+        final Node lastEdgeEndNode = completeWayEdges.get(completeWayEdges.size() - 1).end();
+
+        final Set<Edge> firstEdgeStartNodeEdges = firstEdgeStartNode.connectedEdges();
+        firstEdgeStartNodeEdges
+                .removeIf(edge -> edge.getOsmIdentifier() == edgeBeingVerified.getOsmIdentifier());
+        final Set<Edge> lastEdgeEndNodeEdges = lastEdgeEndNode.connectedEdges();
+        lastEdgeEndNodeEdges
+                .removeIf(edge -> edge.getOsmIdentifier() == edgeBeingVerified.getOsmIdentifier());
+
+        HighwayTag edgeBeingVerifiedHighwayTag = HighwayTag.NO;
+        final Optional<HighwayTag> edgeBeingVerifiedHighwayTagOptional = HighwayTag
+                .highwayTag(edgeBeingVerified);
+        if (edgeBeingVerifiedHighwayTagOptional.isPresent())
+        {
+            edgeBeingVerifiedHighwayTag = edgeBeingVerifiedHighwayTagOptional.get();
+        }
+
+        final Set<HighwayTag> firstEdgeStartNodeEdgesHighwayTags = this
+                .getHighwayTags(firstEdgeStartNodeEdges);
+        final Set<HighwayTag> lastEdgeEndNodeEdgesHighwayTags = this
+                .getHighwayTags(lastEdgeEndNodeEdges);
+
+        // Check ways' first edge's connected edges for suspiciously large highway tag jumps.
+        for (final Edge firstEdgeEdge : firstEdgeStartNodeEdges)
+        {
+            final Optional<HighwayTag> firstEdgeEdgeHighwayTagOptional = HighwayTag
+                    .highwayTag(firstEdgeEdge);
+            if (firstEdgeEdgeHighwayTagOptional.isPresent()
+                    && edgeBeingVerifiedHighwayTagOptional.isPresent()
+                    && !firstEdgeStartNodeEdgesHighwayTags.contains(edgeBeingVerifiedHighwayTag)
+                    && !this.edgeIsRoundaboutOrCircular(firstEdgeEdge))
+            {
+                final HighwayTag firstEdgeEdgeHighwayTag = firstEdgeEdgeHighwayTagOptional.get();
+
+                // All cases
+                if (this.isCaseOne(edgeBeingVerifiedHighwayTag, firstEdgeEdgeHighwayTag)
+                        || this.isCaseTwo(edgeBeingVerifiedHighwayTag, firstEdgeEdgeHighwayTag)
+                        || this.isCaseThree(edgeBeingVerifiedHighwayTag, firstEdgeEdgeHighwayTag))
+                {
+                    markAsFlagged(edgeBeingVerified.getOsmIdentifier());
+                    return Optional.of(this.createFlag(object,
+                            this.getLocalizedInstruction(0, object.getOsmIdentifier())));
+                }
+            }
+        }
+
+        // Check ways' last edge's connected edges for suspiciously large highway tag jumps
+        for (final Edge lastEdgeEdge : lastEdgeEndNodeEdges)
```

Comment:
do you think this loop and the one prior could be put into a method, and parameterized with `firstEdgeStartNodeEdgesHighwayTags` and then `lastEdgeEndNodeEdgesHighwayTags` ?

Reply:
Did my best to implement these in methods. I have it working with same results and eliminates code smell.